lib/widgets/actions.dart

```diff
@@ -21,41 +21,63 @@ Future<bool> markNarrowAsRead(
   Narrow narrow,
   bool useLegacy, // TODO(server-6)
 ) async {
-  try {
-    final store = PerAccountStoreWidget.of(context);
-    final connection = store.connection;
-    if (useLegacy) {
-      await _legacyMarkNarrowAsRead(context, narrow);
-      return true;
-    }
+  if (useLegacy) {
+    await _legacyMarkNarrowAsRead(context, narrow);
+    return true;
+  }
 
-    // Compare web's `mark_all_as_read` in web/src/unread_ops.js
-    // and zulip-mobile's `markAsUnreadFromMessage` in src/action-sheets/index.js .
-    final zulipLocalizations = ZulipLocalizations.of(context);
-    final scaffoldMessenger = ScaffoldMessenger.of(context);
+  // Compare web's `mark_all_as_read` in web/src/unread_ops.js
+  // and zulip-mobile's `markAsUnreadFromMessage` in src/action-sheets/index.js .
+  final zulipLocalizations = ZulipLocalizations.of(context);
+
+  return updateMessageFlagsStartingFromAnchor(
+    context: context,
     // Use [AnchorCode.oldest], because [AnchorCode.firstUnread]
     // will be the oldest non-muted unread message, which would
     // result in muted unreads older than the first unread not
     // being processed.
-    Anchor anchor = AnchorCode.oldest;
-    int responseCount = 0;
-    int updatedCount = 0;
-
+    startAnchor: AnchorCode.oldest,
     // Include `is:unread` in the narrow.  That has a database index, so
     // this can be an important optimization in narrows with a lot of history.
     // The server applies the same optimization within the (deprecated)
     // specialized endpoints for marking messages as read; see
     // `do_mark_stream_messages_as_read` in `zulip:zerver/actions/message_flags.py`.
-    final apiNarrow = narrow.apiEncode()..add(ApiNarrowIsUnread());
+    apiNarrow: narrow.apiEncode()..add(ApiNarrowIsUnread()),
+    includeAnchor: false,
+    op: UpdateMessageFlagsOp.add,
+    flag: MessageFlag.read,
+    onCompletedMessage: zulipLocalizations.markAsReadComplete,
+    onFailedTitle: zulipLocalizations.errorMarkAsReadFailedTitle,
+    // No need to say how many messages, as the [MarkAsUnread] widget
+    // should follow along.
+    onProgressTitle: zulipLocalizations.markAsReadInProgress);
+}
+
+Future<bool> updateMessageFlagsStartingFromAnchor({
+  required BuildContext context,
+  required Anchor startAnchor,
+  required List<ApiNarrowElement> apiNarrow,
+  required bool includeAnchor,
+  required UpdateMessageFlagsOp op,
+  required MessageFlag flag,
+  required String Function(int num) onCompletedMessage,
+  required String onFailedTitle,
+  required String onProgressTitle,
+}) async {
+  try {
+    final store = PerAccountStoreWidget.of(context);
+    final connection = store.connection;
+
+    final zulipLocalizations = ZulipLocalizations.of(context);
```

Comment:
nit: At first I was confused as of why we still need this after having the parameters for the error messages, but then I saw it used for the `errorInvalidResponse` below. We can move `zulipLocalizations` to closer to where it is used, or even just inline it because it is only used once.